validator_worker/src/follower.rs

```diff
@@ -66,44 +65,47 @@ async fn on_new_state<'a, A: Adapter + 'static>(
         return Ok(on_error(&iface, &new_state, InvalidNewState::RootHash).await);
     }
 
-    let adapter = iface
-        .adapter
-        .read()
-        .compat()
-        .await
-        .expect("on_new_state: failed to acquire read lock adapter");
+    let adapter = iface.adapter.read().await;
 
     if !adapter.verify(
         &iface.channel.spec.validators.leader().id,
         &proposed_state_root,
         &new_state.signature,
     )? {
+        drop(adapter);
         return Ok(on_error(&iface, &new_state, InvalidNewState::Signature).await);
     }
 
     let last_approve_response = iface.get_last_approved().await?;
     let prev_balances = last_approve_response
         .last_approved
         .and_then(|last_approved| last_approved.new_state)
-        .map_or(Default::default(), |new_state| new_state.balances);
+        .map_or(Default::default(), |new_state| match new_state.msg {
```

Comment:
isn't this taking quite a bit out of the expressiveness? I.e. instead of having a `new_state` we have the whole message and we need to always match it.

Reply:
Again the response from sentry isn't NewState but a `ValidatorMessage`